Control_Toolkit_ASF/Cost_Functions/CartPole/cartpole_trajectory_cost.py

```diff
@@ -0,0 +1,155 @@
+
+
+    def get_stage_cost(self, states: TensorType, inputs: TensorType, previous_input: TensorType):
+        """
+        Computes costs of all stages, i.e., costs of all rollouts for each timestep.
+
+        :param states: Tensor of states [rollout, timestep, state].
+            rollout dimension is the batch dimension of the rollouts, e.g. 1000 parallel rollouts.
+            timestep dimension are the timesteps of each rollout, e.g. 50 timesteps.
+            state is the state vector, defined by the STATE_INDICES in [state_utilities.py](../CartPole/state_utilities.py)
+        :param inputs: control input, e.g. cart acceleration
+        :param previous_input: previous control input, can be used to compute cost of changing control input
+
+        :returns: Tensor of [rollout, timestamp] over all rollouts and horizon timesteps
+        """
+        # print("test") # only prints once since TF graph is only built once, and not rebuilt if arguments do not change
+
+        dd = 0 # self.dd_weight * self._distance_difference_cost(states[:, :, POSITION_IDX]) # compute cart position target distance cost
+        control_cost = 0 # self.cc_weight * self.R*self._CC_cost(inputs) # compute the cart acceleration control cost
+        control_change_cost = 0 # compute the control change cost
+        ep=0 # pole angle cost
+        position=states[:, :, POSITION_IDX]
+        positionD=states[:, :, POSITIOND_IDX]
+        angle=states[:, :, ANGLE_IDX]
+        angleD=states[:,:,ANGLED_IDX]
+        gui_target_position=self.controller.target_position # GUI slider position
+        gui_target_equilibrium=self.controller.target_equilibrium # GUI switch +1 or -1 to make pole target up or down position
+        input_shape=self.lib.shape(states)
+        num_rollouts=input_shape[0]
+        num_timesteps=input_shape[1]
+        trajectory_cost=self.lib.zeros((num_rollouts,num_timesteps))
+        target_trajectory=self.controller.target_trajectory # [state, timestep]
+        # "angle", "angleD", "angle_cos", "angle_sin", "position", "positionD"
+        cost_weights=(self.pole_angle_weight, self.pole_swing_weight, self.pole_angle_weight, self.pole_angle_weight, self.cart_pos_weight, self.cart_vel_weight)
+        for i in range(NUM_STATES):
+            if not self.lib.any(self.lib.isnan(target_trajectory[i,0])): # to skip a state, the first timestep is NaN
+                state_i=states[:,:,i] # matrix [rollout,timestep] for this state element
+                trajectory_i=target_trajectory[i,:] # timestep vector for this state element
+                zerodiffs=self.lib.zeros((num_rollouts, num_timesteps-1))
+                terminaldiffs=state_i[:,-1]-trajectory_i[-1]
+                terminaldiffs_unsqueezed=self.lib.unsqueeze(terminaldiffs,1)
+                diff=self.lib.concat((zerodiffs,terminaldiffs_unsqueezed),1) # -1 is terminal state, subtracts time vector trajectory_i from each time row i of state
```

Comment:
```suggestion
                diff = state_i - self.lib.unsqueeze(trajectory_i, 0)
```
The only problem there may have been was this:
`state_i` has shape `(320,35)`, while `trajectory_i` has `(35,)`.
In order to take the difference, they have to be castable.
One easy way is to make `trajectory_i` have shape `(1,35)`.